### PR TITLE
login-api

### PR DESCRIPTION
**Pull request by tranvuminhtriet**

### CODE REVIEW

File: src/users/dto/create-user.dto.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  @IsEmail()
+  email: string;
+
+  @IsString()
+  @MinLength(8)
+  @MaxLength(32)
+  @Matches(/((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, {
+    message: 'Password is too weak',
+  })
+  password: string;
```
Comment (tranmanhlinh) [inline src/users/dto/create-user.dto.ts]: Create user should have confirmPassword field.

File: src/users/user.repository.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    const { email, password, firstName, lastName, avatar } = createUserDto;
+
+    const saltRounds = 10;
+    const salt = await bcrypt.genSalt(saltRounds);
+    const hash = await bcrypt.hash(password, salt);
+    // hash password
+    //const salt = await bcrypt.gensalt();
+    //const hashPassword = await bcrypt.hash(password, salt);
+
+    //console.log(salt);
```
Comment (tranmanhlinh) [inline src/users/user.repository.ts]: Remove console.log

File: src/users/user.repository.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import { User } from './user.entity';
+import * as bcrypt from 'bcrypt';
+import {
+  ConflictException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+
+@EntityRepository(User)
+export class UserRepository extends Repository<User> {
+  async createUser(createUserDto: CreateUserDto): Promise<void> {
```
Comment (tranmanhlinh) [inline src/users/user.repository.ts]: Return User entity after create.

File: src/users/user.repository.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      password: hash,
+      firstName,
+      lastName,
+      avatar,
+    });
+
+    try {
+      await this.save(user);
+    } catch (error) {
+      if (error.code === '23505')
```
Comment (tranmanhlinh) [inline src/users/user.repository.ts]: Define the enum for every error code.

File: src/users/users.service.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    @InjectRepository(UserRepository) private userRepositry: UserRepository,
+    private jwtService: JwtService,
+  ) {}
+  async signUp(createUserDto: CreateUserDto): Promise<void> {
+    return this.userRepositry.createUser(createUserDto);
+  }
+
+  async signIn(
+    userCredentialDto: UserCredentialDto,
+  ): Promise<{ accessToken: string }> {
```
Comment (tranmanhlinh) [inline src/users/users.service.ts]: Define DTO for the return object.

File: src/users/users.service.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import { UserRepository } from './user.repository';
+
+@Injectable()
+export class UsersService {
+  constructor(
+    @InjectRepository(UserRepository) private userRepositry: UserRepository,
+    private jwtService: JwtService,
+  ) {}
+  async signUp(createUserDto: CreateUserDto): Promise<void> {
+    return this.userRepositry.createUser(createUserDto);
```
Comment (tranmanhlinh) [inline src/users/users.service.ts:19]: Return UserDto after signUp

File: src/users/dto/user-credential.dto.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+export class UserCredentialDto {
+  @IsEmail()
+  email: string;
+
+  @IsString()
+  @MinLength(8)
+  @MaxLength(32)
+  @Matches(/((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, {
+    message: 'Password is too weak',
```
Comment (tranmanhlinh) [inline src/users/dto/user-credential.dto.ts]: Define error messages as contants.

File: src/users/dto/create-user.dto.ts (hunk 1 above)
Comment (tranvuminhtriet) [reply src/users/dto/create-user.dto.ts]: Done

File: src/users/user.repository.ts (hunk 2 above)
Comment (tranvuminhtriet) [reply src/users/user.repository.ts]: Done

File: src/users/user.repository.ts (hunk 3 above)
Comment (tranvuminhtriet) [reply src/users/user.repository.ts]: Done

File: src/users/user.repository.ts (hunk 4 above)
Comment (tranvuminhtriet) [reply src/users/user.repository.ts]: Done

File: src/users/users.service.ts (hunk 5 above)
Comment (tranvuminhtriet) [reply src/users/users.service.ts]: Done

File: src/users/users.service.ts (hunk 6 above)
Comment (tranvuminhtriet) [reply src/users/users.service.ts:19]: Done

File: src/users/dto/user-credential.dto.ts (hunk 7 above)
Comment (tranvuminhtriet) [reply src/users/dto/user-credential.dto.ts]: Done

**Pull request merged**